### DISCUSSION TITLE
Store Vs useState and constants best practices

### DISCUSSION THREAD

**Discussion by Lozoute** [nuxt/nuxt#16697] [Questions]
Hello all,

I am quite new to nuxt3 (which is awesome by the way) so please excuse me if my questions sounds dumb ^^

### Store Vs UseState

If I am not mistaken, `useState` can replace any store like Pinia? Given that `useState` make ref sharable accross all components I have a hard time knowing if I should use a store or a `useState` ref. For instance, let's say I have a `currentUser` (firstName, lastName, email) that I fetch from a distant API at page loading. Should I stock this information in a Pinia userStore or in a `useState('currentUser')` ?

What are best practices on this topic?

### Constants

Now, let's say I have a `LeftMenu` component with a template like that

```
<ul>
  <li v-for="item in menuItems">{{ item.text }}</li>
</ul>
```

And I have two arrays somewhere in my applications

```
[
  { text: 'foo'  },
  { text: 'bar' },
]
```

```
[
  { text: 'baz'  },
]
```

I want to pass the menuItems as a prop to my `LeftMenu` component.
I know that the content of my arrays does not need to be reactive because it won't change, futhermore I will rarely update those items in my code so it seems like it should be constants.

But I can't figure where I should put those arrays, I thought about a composable with `useState('menuItems1', () => [])` but I'm not sure it is a good way to do it. I also thought about creating a plugin with `provide` so I can access it like that `this.$menuItems1`, but again doesn't feel right to me.

Then I thought about creating a constants folder in my nuxt application and simply importing constants as needed, but again doesn't feel right.

How would you do it? 

Thank you all for your help!

**Comment by Lozoute** [comment]
Hello guys, I didn't have any answer so I am trying a good ol' "up" :'D

Thanks!

**Comment by viii47** [comment]
Hello,

I've asked myself these questions a long while ago around November but I decided against using a third-party plugin like vuex/pinia for Nuxt3 since as you said, `useState` is *quite* the same as `useStore`. Pinia adds some helpers, though, you can easily use them yourself with `useState` as well. There's is no *best practice* for this (in my opinion), more of a personal preference of what you think is best for your current project.

They both work quite equally in terms of setting, getting and mutating data. Example for `useState()` for a login request
```ts
[..composable/user.ts]
export const User = () => {
    const user = useState(
        "user",
        () =>
            ({
                email: "",
                firstname: "",
                lastname: "",
                logged_in: false,
                logged_in_process: "inactive",
                logged_in_status: "",
                register_status: "",
                register_process: "inactive",
                info_process: "inactive",
                status: "",
                addresses: [],
                fetch: true,
            } as IUser)
    );
   [...]
    async function login(mail: string, password: string): Promise<IUser> {
        return await $fetch<IUser>(`${API_USER_LOGIN}`, {
            method: "POST",
            body: {
                email: encodeURI(mail),
                password: encodeURI(password),
            },
        })
    }
    return {
        user,
        login,
    };
};
[..component/login.vue]
[...]
function loginLocal(): void {
    user.value.logged_in_process = "active";
    nextTick(() => {
        User()
            .login(login.email, login.password)
            .then((response) => {
              console.log(response);
                debugLog({
                    origin: "account / login",
                    message: "loginLocal() formData",
                    attachedValue: user,
                });
                if (response.statusCode === 200) {
                    const userHash = useCookie<string | null>("p__h", {
                        path: "/",
                        maxAge: 3600 * 24 * 60,
                    });
                    const userId = useCookie<string | null>("p__i", {
                        path: "/",
                        maxAge: 3600 * 24 * 60,
                    });
                    userHash.value = `${response.hashed_password}`;
                    userId.value = `${response.id}`;
                    user.value.email = response.email;
                    user.value.firstname = response.firstname;
                    user.value.lastname = response.lastname;
                    user.value.logged_in_process = "done";
                    router.push({
                        path: `/${utilGetLocaleBasepath()}/account/overview`,
                    });
                } else {
                    user.value.logged_in_process = "error";
                }
            })
            .catch((error) => {});
    });
}
[...]
```
This isn't by any measure a top-tier example (needs to be rewritten to use the handlers of success/error inside the component)


To your menu example - if they rarely change, then you can simply store that array in the component or page itself. I'd personally not put such static information into a composable - [pinia addresses this question nicely](https://pinia.vuejs.org/getting-started.html#when-should-i-use-a-store). Either you have it inside the component itself (as an array or directly in the template statically), or you have a different folder where you store such info in a json format - which works as well.

**Comment by Lozoute** [comment]
Thanks a lot!

You just confirmed that I can/should use useState instead of a store, I think it is a quite cool way of doing things but I wasn't sure it was "clean" ^^

For the arrays, I get what you're saying but by putting them in the component itself I'm afraid it's gonna get messy.

Let me explain the situation :

I have a LeftMenu component but my admin users can display the app in two ways "Admin mode" or "User mode" it's kinda the same logic as AirBNB, you can either be in "host mode" or "traveler mode". But the menu is the same, same look, same feeling, same everything, only the NuxtLink urls and text will change.

So my logic is that I will use the same component LeftMenu and provide the "admin array" or the "user array" depending of the current choice of my user. But maybe it will be better to have a AdminLeftMenu and a UserLeftMenu? I always have difficulties with those questions because I feel like no answer is perfect

**Comment by husayt** [comment]
Rule of thumb:
- *useState* for just getting and using the data. 
- *pinia/useStore* If you need to have multiple getters or setters on that data

**Reply by ricky11** [reply]
Can i add items to useState from other components ( i know i can mutate, but can i add items to an array?)
